mythril_core/src/acpi/hpet.rs

```diff
@@ -0,0 +1,155 @@
+            sdt.table[offsets::PAGE_PROTECTION] & 0xF,
+        ) {
+            Some(page_protection) => page_protection,
+            None => {
+                return Err(Error::InvalidValue(format!(
+                    "Invalid HPET Page Protection type: {}",
+                    sdt.table[offsets::PAGE_PROTECTION] & 0xF
+                )));
+            }
+        };
+
+        let hardware_rev_id = (event_timer_block_id & 0xFF) as u8;
+        let comparator_count = ((event_timer_block_id & 0x1F00) >> 8) as u8;
+        let counter_cap = ((event_timer_block_id & 0x2000) >> 13) as u8;
+        let legacy_replacement = ((event_timer_block_id & 0x8000) >> 15) as u8;
+        let pci_vendor_id = ((event_timer_block_id & 0xFFFF0000) >> 16) as u16;
+
+        let counter_cap = match counter_cap {
+            0 => false,
+            _ => true,
+        };
+
+        let legacy_replacement = match legacy_replacement {
+            0 => false,
+            _ => true,
+        };
+
+        Ok(Self {
+            sdt,
+            hardware_rev_id,
+            comparator_count,
+            counter_cap,
+            legacy_replacement,
+            pci_vendor_id,
+            address,
+            hpet_number,
+            minimum_tick,
+            page_protection,
+        })
+    }
+}
+
+impl<'a> fmt::Debug for HPET<'a> {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "{:?}", self.sdt)?;
+        write!(f, " HPET address=0x{:x}", self.address.address)
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use super::*;
+    use crate::acpi::{AccessSize, AddressSpaceID};
+
+    #[test]
+    fn test_hpet_parse() {
+        // sample HPET ACPI entry
+        let buf = [
+            0x48, 0x50, 0x45, 0x54, 0x38, 0x00, 0x00, 0x00, 0x01, 0xb6, 0x41,
+            0x4c, 0x41, 0x53, 0x4b, 0x41, 0x41, 0x20, 0x4d, 0x20, 0x49, 0x00,
+            0x00, 0x00, 0x09, 0x20, 0x07, 0x01, 0x41, 0x4d, 0x49, 0x2e, 0x05,
+            0x00, 0x00, 0x00, 0x01, 0xa7, 0x86, 0x80, 0x00, 0x40, 0x00, 0x00,
+            0x00, 0x00, 0xd0, 0xfe, 0x00, 0x00, 0x00, 0x00, 0x00, 0xee, 0x37,
+            0x00,
+        ];
+
+        let hpet_sdt = unsafe { SDT::new(buf.as_ptr()).unwrap() };
+        let hpet = HPET::new(&hpet_sdt).unwrap();
+
+        assert_eq!(hpet.hardware_rev_id, 1);
+        assert_eq!(hpet.comparator_count, 7);
+        assert_eq!(hpet.counter_cap, true);
+        assert_eq!(hpet.legacy_replacement, true);
+        assert_eq!(hpet.pci_vendor_id, 0x8086);
+        assert_eq!(hpet.address.address_space, AddressSpaceID::SystemMemory);
+        assert_eq!(hpet.address.bit_width, 64);
+        assert_eq!(hpet.address.bit_offset, 0);
+        assert_eq!(hpet.address.access_size, AccessSize::Undefined);
+        assert_eq!(hpet.address.address, 0xfed00000);
+        assert_eq!(hpet.hpet_number, 0);
+        assert_eq!(hpet.minimum_tick, 0x37ee);
+        assert_eq!(hpet.page_protection, PageProtection::NoProtection);
```

Comment:
How did you determine these were the correct values?

Reply:
I used the struct definitions from https://wiki.osdev.org/HPET and created a simple C program to read in the HPET entry on my machine, cast it to a `struct hpet`, and print out the values. I then copied the HPET entry on my machine into the byte array used for the test.